feat: Port PoiCreationScreen to MapData

ShadowCat117 opened the pull request:
![2024-11-05_11 14 17](https://github.com/user-attachments/assets/dc4ae197-fa56-4b9f-8fe2-204559dbe280)
![2024-11-05_11 14 29](https://github.com/user-attachments/assets/cf951307-f793-4d7e-ac19-95eec2fc4388)
![2024-11-05_11 14 39](https://github.com/user-attachments/assets/34ccef43-5288-4ac4-90b1-59097068eb80)
![2024-11-05_11 15 03](https://github.com/user-attachments/assets/9fa461eb-dbf6-4f45-a81a-59bd55fa75db)
![2024-11-05_11 15 30](https://github.com/user-attachments/assets/8b426053-422e-4635-83b3-03cf1e56459f)
![2024-11-05_11 16 08](https://github.com/user-attachments/assets/0726ba1d-670f-4040-9f58-1a810a37b49b)
![2024-11-05_11 16 32](https://github.com/user-attachments/assets/2140b225-a094-4fd1-92b3-390640a2d3c2)

ShadowCat117 committed:
chore: Rename PoiCreationScreen to WaypointCreationScreen

ShadowCat117 committed:
chore: Rename i18n key

ShadowCat117 committed:
wip: Port WaypointCreationScreen to MapData

ShadowCat117 committed:
feat: Add title to creation screen

ShadowCat117 committed:
fix: Hide all icon widgets if not using icon on reinit

ShadowCat117 committed:
fix: Fix text rendering on category widgets

ShadowCat117 committed:
feat: Category screen rendering improvements

ShadowCat117 committed:
feat: Custom icon screen improvements

ShadowCat117 committed:
fix: Set icon ID to none after toggling icon off

ShadowCat117 committed:
fix: Set icon ID after toggling icon back on if selected

ShadowCat117 committed:
feat: Add always/never visibility buttons

ShadowCat117 committed:
chore: Uncomment checkInvalidAttribute

ShadowCat117 commented on common/src/main/java/com/wynntils/services/mapdata/attributes/MapAttributesBuilder.java:
```diff
@@ -106,17 +108,17 @@ public MapAreaAttributesBuilder asAreaAttributes() {
 
     protected void checkInvalidAttribute(String fieldName) {
         // FIXME: WaypointLocation is giving Unsupported attribute set: borderWidth when using WaypointCreationScreen
```
If you want to try out this pr, you will need to comment out this method. Fix can either be pushed to this branch or the main MapData one and will just update this branch to have the fix

kristofbolyai replied:
It's fixed in my latest mapdata pr.

kristofbolyai replied:
Can you remove this FIXME now that it is fixed upstream?

ShadowCat117 committed:
ci: spotless formatting

kristofbolyai committed:
fix: Make MapXAttributesImpl have custom deserializers instead of MapFeatures to avoid type issues

This is the best of two worlds, now the MapFeatureImpl classes can be extended safely whenever we want a serializable/deserializable type, without having to worry that their type is always MapXImpl (as that is the best we could do with GSON's type hierarchy adapters). With this approach, we get to safely and correctly deserialize attributes.

This change does remove the old feature deserializer's featureId -> id and categoryId -> category mapping, which I actually find to be okay, as now the json representation is even closer to the classes themselves.

kristofbolyai committed:
Merge branch 'mapdata-serialization-fix' into waypoint-creation-mapdata

# Conflicts:
#	common/src/main/java/com/wynntils/core/json/JsonManager.java
#	common/src/main/java/com/wynntils/services/mapdata/providers/json/JsonProvider.java

kristofbolyai committed:
fix: Fix serialization

kristofbolyai committed:
fix: Add attribute serializer

kristofbolyai committed:
Merge branch 'mapdata-serialization-fix' into waypoint-creation-mapdata

kristofbolyai committed:
Merge branch 'mapdata-ports' into waypoint-creation-mapdata

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/WaypointCreationScreen.java:
```diff
@@ -0,0 +1,940 @@
+                .setIconColor(iconColorCache)
+                .setIconVisibility(iconVisibility)
+                .asLocationAttributes()
+                .build();
```
Can you take a look at this? I added this comment on some other branch, but it is important here.

ShadowCat117 replied:
Does this only apply to visibility? As far as I can tell, the others have a valid default value, except for location and label which are required to be input

kristofbolyai replied:
If no visibility is set here, null values won't be serialized with the new serializer.

kristofbolyai commented:
Before going through the whole code, let's talk functionality. There is a slight issue with the icon - label visibility not being clear, so we need to signal to the user that there is no default value, or if there is one set by the user, what it is. As default label visibility for personal waypoints is never, and as this UI sets no default icons, it is easy to create waypoints on the map that never show up. For the category, it would be nice to signal to the user that whatever they will set is going to be a subcategory under `"wynntils:personal:waypoint"`.

kristofbolyai commented:
However, I have to say, I much prefer this UI to the old one, in terms of functionality. Great!

kristofbolyai commented:
Reminder for me to fix MapDataUtils#sanitize :)

ShadowCat117 committed:
feat: Display full category to user

ShadowCat117 committed:
feat: Make flag icon the default

ShadowCat117 committed:
feat: Set default values for visibility. Display visibility levels under buttons

kristofbolyai commented on common/src/main/java/com/wynntils/services/mapdata/WaypointsService.java:
```diff
@@ -40,12 +47,36 @@ public void onStorageLoad(Storage<?> storage) {
+    }
+
+    public Set<String> getCategories() {
+        return getWaypoints().stream().map(MapLocationImpl::getCategoryId).collect(Collectors.toSet());
```
```suggestion
        return getWaypoints().stream().map(MapLocationImpl::getCategoryId).collect(Collectors.toUnmodifiableSet());
```

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/CustomIconWidget.java:
```diff
@@ -0,0 +1,117 @@
+        FontRenderer.getInstance()
+                .renderScrollingText(
+                        poseStack,
+                        StyledText.fromString(customIcon.getIconId().substring("wynntils:icon:personal:".length())),
```
Honestly, the thing I don't like about mapdata are these "floating" strings referencing categories. For now, make this a constant at least.

kristofbolyai reviewed (approved):
Essentially this looks good, please finish the reviews I've added yesterday.

ShadowCat117 committed:
chore: Remove fixme comment

ShadowCat117 committed:
chore: Make categories return unmodifiable

ShadowCat117 committed:
chore: Make icon id prefix a constant

kristofbolyai committed:
Merge branch 'mapdata-ports' into waypoint-creation-mapdata

# Conflicts:
#	common/src/main/java/com/wynntils/screens/maps/PoiCreationScreen.java
#	common/src/main/resources/assets/wynntils/lang/de_de.json
#	common/src/main/resources/assets/wynntils/lang/en_us.json
#	common/src/main/resources/assets/wynntils/lang/es_es.json
#	common/src/main/resources/assets/wynntils/lang/fr_fr.json
#	common/src/main/resources/assets/wynntils/lang/hu_hu.json
#	common/src/main/resources/assets/wynntils/lang/it_it.json
#	common/src/main/resources/assets/wynntils/lang/ja_jp.json
#	common/src/main/resources/assets/wynntils/lang/ko_kr.json
#	common/src/main/resources/assets/wynntils/lang/nl_nl.json
#	common/src/main/resources/assets/wynntils/lang/pl_pl.json
#	common/src/main/resources/assets/wynntils/lang/pt_pt.json
#	common/src/main/resources/assets/wynntils/lang/ru_ru.json
#	common/src/main/resources/assets/wynntils/lang/sv_se.json
#	common/src/main/resources/assets/wynntils/lang/zh_cn.json
#	common/src/main/resources/assets/wynntils/lang/zh_tw.json

kristofbolyai committed:
fix: Port to 1.21.4

magicus commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -63,7 +63,8 @@ public PoiManagerWidget(
         editButton = new Button.Builder(
                         Component.translatable("screens.wynntils.poiManagementGui.edit"),
-                        (button) -> McUtils.mc().setScreen(PoiCreationScreen.create(managementScreen, poi)))
+                        // FIXME: null should be a WaypointLocation object (fix when porting PoiManagementScreen)
```
I'm probably reading all the PRs in the wrong order, but this FIXME seems to have been checked in to the mapdata branch. Has it been addressed in a later PR or is it still pending?

ShadowCat117 replied:
Fixed in https://github.com/Wynntils/Wynntils/pull/3123
